Src/Common/InstrumentationKeyHashLookupHelper.cs

```diff
@@ -0,0 +1,55 @@
+﻿namespace Microsoft.ApplicationInsights.Common
+{
+    using System.Collections.Concurrent;
+    using System.Globalization;
+    using System.Security.Cryptography;
+    using System.Text;
+
+    /// <summary>
+    /// A store for instrumentation key hashes. This is order so as to optimize the computation of hashes.
+    /// </summary>
+    public static class InstrumentationKeyHashLookupHelper
+    {
+        private static ConcurrentDictionary<string, string> knownIKeyHashes = new ConcurrentDictionary<string, string>();
+
+        /// <summary>
+        /// Retrieves the hash for a given instrumentation key.
+        /// </summary>
+        /// <param name="instrumentationKey">Instrumentation key string.</param>
+        /// <returns>SHA256 hash for the instrumentation key.</returns>
+        public static string GetInstrumentationKeyHash(string instrumentationKey)
+        {
+            if (string.IsNullOrEmpty(instrumentationKey))
+            {
+                return null;
+            }
+
+            if (!knownIKeyHashes.ContainsKey(instrumentationKey))
+            {
+                knownIKeyHashes[instrumentationKey] = GenerateSHA256Hash(instrumentationKey);
```

Comment:
Constant or a variable?  I agree with returning a variable from here rather than falling through to the return below.